src/OrchardCore/OrchardCore.Infrastructure/Security/OrchardCoreBuilderExtensions.cs

```diff
@@ -30,6 +33,31 @@ public static OrchardCoreBuilder AddSecurity(this OrchardCoreBuilder builder)
             services.AddScoped<IAuthorizationHandler, PermissionHandler>();
         });
 
+        builder.Configure(ValidatePermissionsAsync);
+
         return builder;
     }
+
+    private static async ValueTask ValidatePermissionsAsync(IApplicationBuilder builder, IEndpointRouteBuilder routeBuilder, IServiceProvider serviceProvider)
+    {
+        // Make sure registered permissions are valid, i.e. they must be unique.
+        var permissionProviders = serviceProvider.GetServices<IPermissionProvider>();
```

Comment:
This logic is repeated few times. I think it would be helpful to add a new method to `DefaultPermissionService` that would keep track of the duplicated permission.

This way this is something we don't need to compute more than once during the tenant lifecycle.

Reply:
As the permissions are usually only required in the admin portal, I don't think it is good to cache them forever. And adding them to the memory cache is also not really required, because computing them is cheap enough.

Regarding adding this to the permission service instead, it would also mean to log this error in the service each time the permissions are enumerated. I therefore think this one duplication of similar code is a better choice.

Reply:
We already keep track of all the permissions in that singleton class. In the LoadPermission method which is computed once, you can also check for duplication and store duplicated permissions. This way you can log more info as the full name (including namespace) of the conflicted permissions.  I don't think it would take up much memory since most of the time we won't have the duplicate issue and if we did, it won't be much. So I think it's better to cache duplication "if any", this way it's always available and we don't have to compute it more than once and it is part of a service instead of doing the same check is multiple places.

Reply:
Oh, you're right. I'll think more about this...

Reply:
Actually you were wrong ;-) The `DefaultPermissionService` is registered as scoped and the permissions are not cached by it for the app lifetime. And they are not required to be cached forever as well, as they are used only in the portal and even there not all the time. So I don't think we should change that.

I also don't think it is a good idea to add the "duplicated permissions" info as a new public API surface to the `IPermissionService`. This is a internal implementation detail, that should not be exposed publicly.

I therefore propose to leave it separated as suggested.

Reply:
Okay. I for some reason thought DefaultPermissionService was a singleton object.